### PR TITLE
Allow exclusion of generic interfaces of events with an annotation

### PR DESCRIPTION
**Pull request by liach**
[**SpongeAPI**](https://github.com/SpongePowered/SpongeAPI/pull/1759) | **event-impl-gen**

This is meant to serve abstract event interfaces which does not represent a concrete event.

### CODE REVIEW

**Comment by liach** [comment]
wait a sec on this

**Comment by liach** [comment]
Now it's fixed. Spoon is just weird.

**Comment by ryantheleach** [comment]
@liach I'm going to assume by your comments that this is ready for review. ~qa

**Comment by liach** [comment]
It is indeed.

File: src/main/java/org/spongepowered/eventimplgen/EventInterfaceProcessor.java [hunk 1]
```diff
@@ -61,6 +61,6 @@ public void process(CtInterface<?> event) {
     }
 
     public static boolean shouldGenerate(CtInterface<?> candidate) {
-        return candidate.getNestedTypes().isEmpty() || EventImplGenTask.getAnnotation(candidate, "org.spongepowered.api.util.annotation.eventgen.GenerateFactoryMethod") != null;
+        return (candidate.getNestedTypes().isEmpty() && EventImplGenTask.getAnnotation(candidate, "org.spongepowered.api.util.annotation.eventgen.AbstractEvent") == null) || EventImplGenTask.getAnnotation(candidate, "org.spongepowered.api.util.annotation.eventgen.GenerateFactoryMethod") != null;
```
Comment (stephan-gh) [inline src/main/java/org/spongepowered/eventimplgen/EventInterfaceProcessor.java]: Can you either split this into multiple if statements or at least add some line breaks here? There is way too much scrolling required here and it doesn't quite make it easier to understand :)

**Review by stephan-gh — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 2]
```diff
@@ -33,11 +33,11 @@ configurations {
 }
 
 dependencies {
-    shade 'fr.inria.gforge.spoon:spoon-core:5.7.0'
+    shade 'fr.inria.gforge.spoon:spoon-core:5.5.0'
```
Comment (Aaron1011) [inline build.gradle]: Is there a specific reason for why you downgraded the Spoon dependency?

**Review by Aaron1011 — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 2 above)
Comment (liach) [reply build.gradle]: Because spoon 5.7.0 is broken. The current version of event-impl-gen that sponge api uses is the latest release instead of the snapshot on maven so the issue was never revealed until I tested.

Comment (parlough) [reply build.gradle]: Has the spoon issue been fixed since 5.7 now that they've had quite a few releases since then? What even was the 🥄 issue?

Comment (liach) [reply build.gradle]: Nope, they simply changed the behavior. Unfortunately, I do not understand spoon well enough to update code in event impl gen.

Comment (ryantheleach) [reply build.gradle]: I'm not sure I understand.

> The current version of event-impl-gen that sponge api uses is the latest release instead of the snapshot

But it's clearly specified here at 5.7.0 ? Why would it use the latest instead of the snapshot? Why would we **want** to use the snapshot over the version listed? and 5.5.0 isn't the snapshot?

What behavior changed?

File: build.gradle [hunk 3]
```diff
@@ -33,11 +33,11 @@ configurations {
...
     compile 'org.ow2.asm:asm:5.2'
     compile 'com.google.guava:guava:21.0'
     compileOnly 'com.google.errorprone:error_prone_annotations:2.0.15'
-    compileOnly 'com.google.code.findbugs:jsr305:3.0.1'
+    compile 'com.google.code.findbugs:jsr305:3.0.1'
```
Comment (ryantheleach) [inline build.gradle]: Was this change necessary?

Comment (liach) [reply build.gradle]: This cause issues in code as the code references the `@Nullable` or `@Nonnull` class when it checks for the annotations in source.

File: build.gradle (hunk 2 above)
Comment (liach) [reply build.gradle]: @ryantheleach I mean this: https://github.com/SpongePowered/SpongeAPI/blob/bleeding/build.gradle#L14

File: build.gradle (hunk 3 above)
Comment (liach) [reply build.gradle]: @ryantheleach More exactly, these annotations are used at runtime here: https://github.com/SpongePowered/event-impl-gen/blob/0b10b0fac267f1809c002f983def01339729f7ac/src/main/java/org/spongepowered/eventimplgen/factory/ClassGenerator.java#L168-L174

**Comment by stephan-gh** [comment]
@liach Can you fix the conflicts here? I went ahead and fixed up the dependencies in a separate commit to fix the builds, can just drop all your changes in `build.gradle`.

Also, what's the status of this PR? Is this properly working together with the SpongeAPI changes?

**Comment by liach** [comment]
this change must be applied first. only when this pr is merged and api updates its dependency of event impl gen can the api pr pass the build. i can work on this pr tomorrow morning.

**Comment by gabizou** [comment]
Just because I want to get some of this more stable and expandable for more than just SpongeAPI usage, can this be merged sooner than later so that configuration changes can take place without having pending changes waiting and then re-broken?

**Comment by liach** [comment]
Hmm, I believe I've resolved the requested changes. Any news on this?

**Comment by liach** [comment]
@Aaron1011 Updated to include #7 as well.

**Review by stephan-gh — APPROVED** [review]

**Comment by liach** [comment]
@Aaron1011 Can you please review this again?
